sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,448 @@
+
+            .. note:: The data will be converted to C ordering,
+                which will cause a memory copy
+                if the given data is not C-contiguous.
+
+        y : Ignored
+            Not used, present here for API consistency by convention.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            The weights for each observation in X. If None, all observations
+            are assigned equal weight.
+
+        Returns
+        -------
+        self
+            Fitted estimator.
+        """
+        X = self._validate_data(X, accept_sparse='csr',
+                                dtype=[np.float64, np.float32],
+                                order='C', copy=self.copy_x,
+                                accept_large_sparse=False)
+
+        self._check_params(X)
+        random_state = check_random_state(self.random_state)
+        sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+        self._n_threads = _openmp_effective_n_threads()
+
+        if self._algorithm == "full":
+            self._kmeans_single = _kmeans_single_lloyd
+            self._check_mkl_vcomp(X, X.shape[0])
+        else:
+            self._kmeans_single = _kmeans_single_elkan
+
+        data_left = X
+        weights_left = sample_weight
+
+        centroids = []
+
+        if self.verbose:
+            print("Running Bisecting K-Means with parameters:")
+            print("-> number of clusters: {}".format(self.n_clusters))
+            print("-> number of centroid initializations: {}"
+                  .format(self.n_init))
+            print("-> relative tolerance: {:.4e} \n".format(self.tol))
+
+        for n_iter in range(self.n_clusters):
+
+            # Perform Bisection
+            centers, labels, _,  _ = self._bisect(data_left, y, weights_left,
+                                                  random_state)
+
+            # Check SSE of each of computed centroids
+            errors = self._calc_bisect_errors(X, centers, labels)
+
+            lower_sse_index = 0 if \
+                errors[0].sum(axis=0) < errors[1].sum(axis=0) else 1
+
+            higher_sse_index = 1 if lower_sse_index == 0 else 0
+
+            # Centroid with lower SSE is added to result centroids.
+            centroids.append(centers[lower_sse_index])
+
+            if self.verbose:
+                print("Centroid Found: {}".format(centers[lower_sse_index]))
+
+            if len(centroids) + 1 == self.n_clusters:
+                # Desired number of cluster centers is reached so centroid with
+                # higher SSE would not be split further
+                centroids.append(centers[higher_sse_index])
+
+                if self.verbose:
+                    print("Centroid Found: {}".format(
+                        centers[higher_sse_index], ))
+                break
+
+            # Data with label that has higher SSE will be further processed
+            indexes = [x for x in range(len(labels))
```

Comment:
This is hard to read and inefficient. You can use `indexes = np.flatnonzero(labels == higher_sse_index)`. Or even faster, using a boolean mask `mask = labels == higher_sse_index` and using the mask with `data_left = data_left[mask]`.

Reply:
I firstly used mask in the way you implemented, but was confused with 'performance comparison' shown between those two. Guess It was some issue of measuring time on VM.
I'll change that